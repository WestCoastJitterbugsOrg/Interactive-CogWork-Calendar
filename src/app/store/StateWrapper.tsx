import { canStoreSelection } from '../services/cookies';
import StateContext, { CategoryStore } from './model';
import { createContext, useEffect, useState } from 'react';
import type { WCJ } from 'types';
import type { MaybeArray } from 'types/utils';

export const stateContext = createContext<StateContext>({
	categories: {},
	events: {},
	eventModal: undefined,
});

export type CategoryMap = Readonly<Record<string, CategoryStore>>;
export type EventMap = Readonly<Record<string, Readonly<WCJ.Event>>>;

type Props = {
	children: MaybeArray<JSX.Element>;
	categories: CategoryMap;
	events: EventMap;
	eventModal?: string;
};

export function StateWrapper(props: Props) {
	const [categories, setCategories] = useState<Record<string, CategoryStore>>(
		props.categories
	);
	const [events, setEvents] = useState<EventMap>(props.events);
	const [eventModal, setEventModal] = useState<string | undefined>(
		props.eventModal
	);

	useEffect(() => {
		if (canStoreSelection()) {
			const uncheckedEvents = Object.values(events)
				.filter((x) => !x.showInCalendar)
				.map((x) => x.id);

			localStorage.setItem('uncheckedEvents', JSON.stringify(uncheckedEvents));
		}
	}, [events]);

	return (
		<stateContext.Provider
			value={{
				categories,
				setCategories,
				events,
				setEvents,
				eventModal,
				setEventModal,
			}}
		>
			{props.children}
		</stateContext.Provider>
	);
}
