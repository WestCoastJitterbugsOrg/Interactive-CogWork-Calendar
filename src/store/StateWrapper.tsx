import { createContext, Dispatch, useReducer } from "react";
import { EventActions } from "./event-actions";
import EventStore from "./model";
import eventReducer from "./reducers";

export interface StateContext {
  state: EventStore;
  dispatch: Dispatch<EventActions>;
}
const emptyContext: EventStore = {
  categories: { byId: {}, allIds: [] },
  events: { byId: {}, allIds: [] },
  eventModal: false,
};

export const StateContext = createContext<StateContext>({
  state: emptyContext,
  dispatch: () => null,
});

interface Props {
  children: JSX.Element[];
  initialContext: EventStore;
}

function StateWrapper(props: Props) {
  const [state, dispatch] = useReducer(eventReducer, props.initialContext);

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {props.children}
    </StateContext.Provider>
  );
}

export default StateWrapper;
