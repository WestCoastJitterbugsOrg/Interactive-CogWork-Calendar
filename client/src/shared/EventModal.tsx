import { useContext } from "react";
import Modal from "react-modal";
import { StateContext } from "../App";
import { EventActionTypes } from "../store/reducers";
import { LinkButton } from "./Button";

const customStyles: Modal.Styles = {
  content: {
    top: "50%",
    left: "50%",
    right: "auto",
    bottom: "auto",
    marginRight: "-50%",
    transform: "translate(-50%, -50%)",
    maxWidth: 800,
    maxHeight: "max(80vh, 200px)",
    overflow: "auto",
  },
  overlay: {
    zIndex: 10000,
    backgroundColor: "rgba(0,0,0,75%)",
  },
};

Modal.setAppElement("#wcjcal");

function displayDate(date: Date) {
  return date.toLocaleString("en-GB", {
    dateStyle: "medium",
    timeStyle: "short",
    timeZone: "UTC",
  });
}

export function EventSeriesModal(props: { ref?: HTMLElement }) {
  const { state, dispatch } = useContext(StateContext);

  return (
    <Modal
      onRequestClose={() => {
        dispatch({
          type: EventActionTypes.modalClosed,
        });
      }}
      isOpen={!!state.eventModal}
      style={customStyles}
      parentSelector={() => document.getElementById("wcjcal") || document.body}
    >
      {state.eventModal && (
        <EventSeriesModalContent
          event={state.events.byId[state.eventModal]}
          onCloseClick={() => {
            dispatch({ type: EventActionTypes.modalClosed });
          }}
        ></EventSeriesModalContent>
      )}
    </Modal>
  );
}

function EventSeriesModalContent({
  event,
  onCloseClick,
}: {
  event: Wcj.Event;
  onCloseClick: () => void;
}) {
  const { first, last } = event.occasions.reduce(
    ({ first, last }, curr) => ({
      first: Math.min(first, curr.start.getTime()),
      last: Math.max(last, curr.end.getTime()),
    }),
    {
      first: Number.MAX_SAFE_INTEGER,
      last: Number.MIN_SAFE_INTEGER,
    }
  );

  return (
    <div className="modal-content">
      <div
        style={{ position: "absolute", right: 20, top: 20, cursor: "pointer" }}
        onClick={onCloseClick}
      >
        âœ–
      </div>
      <h4 className="mt-0">{event.title}</h4>
      {event.description?.includes("<p>") ? (
        <div dangerouslySetInnerHTML={{ __html: event.description }} />
      ) : (
        <p>{event.description}</p>
      )}
      <div className="flex items-end">
        <div className="flex-grow">
          <div>
            <strong>Where:</strong> {event.place}
          </div>
          <div>
            <strong>Price:</strong> {event.price}
          </div>
          <div>
            <strong>Instructors:</strong> {event.instructors}
          </div>
          <div>
            <strong>First occasion starts:</strong>{" "}
            {displayDate(new Date(first))}
          </div>
          <div>
            <strong>Last occasion ends:</strong> {displayDate(new Date(last))}
          </div>
        </div>
        <LinkButton
          title="Registration"
          href={event.registrationUrl}
          size="lg"
        />
      </div>
    </div>
  );
}
