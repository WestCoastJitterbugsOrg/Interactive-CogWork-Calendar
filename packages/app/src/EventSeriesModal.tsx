import style from './EventSeriesModal.module.scss';
import closeIcon from './assets/close.svg';
import { stateContext } from './store/StateWrapper';
import { useContext, useEffect, useRef } from 'react';
import * as Modal from 'react-modal';

type Props = {
	parent: HTMLElement;
};

const rootHtmlElement = document.documentElement;

export function EventSeriesModal(props: Props) {
	const { events, eventModal, setEventModal } = useContext(stateContext);
	const event = eventModal && events[eventModal];

	const previousOverflowStyle = useRef<string>();
	useEffect(() => {
		// Code to prevent scrolling events getting triggered on the main page:
		if (eventModal) {
			previousOverflowStyle.current = rootHtmlElement.style.overflow;
			rootHtmlElement.style.overflow = 'hidden';
		} else if (previousOverflowStyle.current) {
			// When modal is closed, we either have to reapply the style that was before
			rootHtmlElement.style.overflow = previousOverflowStyle.current;
		} else {
			// Or remove it if there was none
			rootHtmlElement.style.removeProperty('overflow');
		}
	}, [eventModal]);

	return (
		<Modal
			onRequestClose={() => setEventModal?.(undefined)}
			preventScroll={true}
			isOpen={!!eventModal}
			className={style.modal}
			overlayClassName={style.overlay}
			parentSelector={() => props.parent}
			appElement={props.parent}
		>
			{event && (
				<div className={style.content}>
					<button
						type="button"
						className={style.closeButton}
						onClick={() => setEventModal?.(undefined)}
						aria-label="modal-close-button"
					>
						<img alt="Close" src={closeIcon} width="16" height="16" />
					</button>
					<iframe
						className={style.iframe}
						src={event.registrationUrl}
						title={event.title}
					></iframe>
				</div>
			)}
		</Modal>
	);
}
