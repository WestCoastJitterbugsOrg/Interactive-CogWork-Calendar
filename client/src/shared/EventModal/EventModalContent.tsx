import { EventInfoTable, LinkButton } from "@app/shared";

export function EventSeriesModalContent({
  event,
  onCloseClick,
}: {
  event: Wcj.Event;
  onCloseClick: () => void;
}) {
  return (
    <div className="md:min-w-[700px]">
      <div
        className="absolute right-5 top-5 cursor-pointer"
        onClick={onCloseClick}
      >
        âœ–
      </div>
      <h4 className="mt-0">{event.title}</h4>
      {event.description?.includes("<p>") ? (
        <div dangerouslySetInnerHTML={{ __html: event.description }} />
      ) : (
        <p>{event.description}</p>
      )}
      <div className="flex flex-wrap justify-around gap-4 items-end mt-4">
        <div className="bg-wcj-mint p-4 rounded-md">
          <EventInfoTable event={event}></EventInfoTable>
        </div>
        <div>
          <LinkButton href={event.registrationUrl} size="lg">
            Registration
          </LinkButton>
        </div>
      </div>
    </div>
  );
}
