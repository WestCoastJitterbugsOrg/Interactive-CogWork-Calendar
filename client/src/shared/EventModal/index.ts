import { EventSeriesModal } from "./EventSeriesModal";

export default EventSeriesModal;
