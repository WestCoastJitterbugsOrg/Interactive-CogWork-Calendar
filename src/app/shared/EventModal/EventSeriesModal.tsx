import { appContainer } from '../../../app-container';
import { stateContext } from '../../store/StateWrapper';
import { useContext, useEffect, useRef } from 'react';
import * as Modal from 'react-modal';
import { EventSeriesModalContent } from './EventModalContent';
import { default as style } from './EventSeriesModal.module.scss';

type Props = {
	parent?: HTMLElement;
};

const rootHtmlElement = document.documentElement;

export function EventSeriesModal(props: Props) {
	const { eventModal, setEventModal } = useContext(stateContext);
	const previousOverflowStyle = useRef<string>();
	useEffect(() => {
		// Code to prevent scrolling events getting triggered on the main page:
		if (eventModal) {
			previousOverflowStyle.current = rootHtmlElement.style.overflow;
			rootHtmlElement.style.overflow = 'hidden';
		} else if (previousOverflowStyle.current) {
			// When modal is closed, we either have to reapply the style that was before
			rootHtmlElement.style.overflow = previousOverflowStyle.current;
		} else {
			// Or remove it if there was none
			rootHtmlElement.style.removeProperty('overflow');
		}
	}, [eventModal]);

	return (
		<Modal
			onRequestClose={() => setEventModal?.(undefined)}
			preventScroll={true}
			isOpen={!!eventModal}
			className={style.modal}
			overlayClassName={style.overlay}
			parentSelector={() => props.parent ?? appContainer}
			appElement={props.parent ?? appContainer}
		>
			{eventModal && <EventSeriesModalContent eventId={eventModal} />}
		</Modal>
	);
}
