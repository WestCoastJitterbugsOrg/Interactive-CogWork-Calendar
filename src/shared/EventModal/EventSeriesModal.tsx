import { appContainer } from "@app/app-container";
import { StateContext } from "@app/store/StateWrapper";
import { useContext } from "react";
import Modal from "react-modal";
import { EventSeriesModalContent } from "./EventModalContent";

interface Props {
  parent?: HTMLElement;
}

export function EventSeriesModal(props: Props) {
  const { eventModal, setEventModal } = useContext(StateContext);

  return (
    <Modal
      onRequestClose={() => setEventModal?.(false)}
      preventScroll={true}
      isOpen={!!eventModal}
      className={`
        absolute top-1/2 left-1/2 right-auto bottom-auto -mr-[50%]
        h-[max(80vh,200px)] max-w-[900px] -translate-x-1/2 -translate-y-1/2 
        rounded-md bg-dark pt-12 outline-none`}
      overlayClassName="fixed z-50 inset-0 bg-black bg-opacity-75"
      parentSelector={() => props.parent ?? appContainer}
      appElement={props.parent ?? document.body ?? appContainer}
    >
      {eventModal && <EventSeriesModalContent eventId={eventModal} />}
    </Modal>
  );
}
