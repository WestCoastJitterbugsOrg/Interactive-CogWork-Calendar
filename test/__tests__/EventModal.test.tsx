import "@testing-library/jest-dom";
import { unmountComponentAtNode } from "react-dom";
import { mockStore } from "../__mocks__/stateContext";
import { createRenderer } from "../test-utils";
import StateWrapper, { StateContext } from "@app/store/StateWrapper";
import { EventSeriesModal } from "@app/shared/EventModal/EventSeriesModal";
import { useContext, useEffect } from "react";
import { act } from "react-dom/test-utils";

const renderer = createRenderer();

beforeEach(() => {
  renderer.container = document.createElement("div");
  renderer.container.id = "wcjcal";
  document.body.appendChild(renderer.container);
});

afterEach(() => {
  if (renderer.container != null) {
    unmountComponentAtNode(renderer.container);
    renderer.container.remove();
  }
  renderer.container = null;
});

it("EventModal Snapshot", async () => {
  const ModalWrapper = () => {
    const { setEventModal } = useContext(StateContext);
    useEffect(() => {
      act(() => {
        setEventModal?.(Object.values(mockStore.events)[0].id);
      });
    }, [setEventModal]);
    return <EventSeriesModal />;
  };

  renderer.render(
    <StateWrapper categories={mockStore.categories} events={mockStore.events}>
      <ModalWrapper />
    </StateWrapper>
  );

  expect(renderer.container?.innerHTML).toMatchSnapshot();
});
