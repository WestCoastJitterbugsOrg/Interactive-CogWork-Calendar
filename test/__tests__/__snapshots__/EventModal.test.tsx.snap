// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventModal Snapshot 1`] = `
<body
  aria-hidden="true"
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay fixed z-50 inset-0 bg-black bg-opacity-75"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content 
        absolute top-1/2 left-1/2 right-auto bottom-auto -mr-[50%]
        h-[max(80vh,200px)] max-w-[900px] -translate-x-1/2 -translate-y-1/2 
        rounded-md bg-white pt-12 outline-none"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="h-full w-[min(900px,100vw)]"
          data-testid="event-series-modal-content"
        >
          <div
            class="absolute right-5 top-5 cursor-pointer"
            data-testid="modal-close-button"
          >
            <img
              height="16"
              src="test-file-stub"
              width="16"
            />
          </div>
          <iframe
            class="h-full w-full overflow-hidden border-none outline-none"
            src="https://example.com/registration"
            title="Grundkurs Lindy Hop"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
